gobblin-utility/src/main/java/org/apache/gobblin/util/filesystem/PathAlterationObserver.java

```diff
@@ -43,9 +43,9 @@ public class PathAlterationObserver {
   private final PathFilter pathFilter;
   private final Comparator<Path> comparator;
   private final FileSystem fs;
-
   private final Path[] EMPTY_PATH_ARRAY = new Path[0];
 
+  private boolean changeApplied = false;
```

Comment:
since only `checkAndNotify` is `synchronized`, not the others setting this, it should be `volatile`.

that said, the "check-then-act" of `if (this.changeApplied) {...}` may  be unsafe until those other methods actually are `synchronized`.

Reply:
I'll change the other functions to be synchronized as well, but I was under the impression it would be safe since those other functions are dependent on checkAndNotify, and since only 1 function execution can run at once, then the others shouldn't be able to modify the check at the same time?

Reply:
that's true: if all the others are called only within `checkThenNotify`, which is `synchronized`, then it's not necessary to synchronize them in addition.  I hadn't grasped earlier whether that's the case.